src/navsat_transform.cpp

```diff
@@ -110,66 +118,68 @@ void NavSatTransform::run()
       datum_yaw = datum_vals[2];
     }
 
-    robot_localization::srv::set_datum::request request;
-    request.geo_pose.position.latitude = datum_lat;
-    request.geo_pose.position.longitude = datum_lon;
-    request.geo_pose.position.altitude = 0.0;
+    auto request = std::make_shared<robot_localization::srv::SetDatum::Request>();
+    request->geo_pose.position.latitude = datum_lat;
+    request->geo_pose.position.longitude = datum_lon;
+    request->geo_pose.position.altitude = 0.0;
     tf2::Quaternion quat;
     quat.setRPY(0.0, 0.0, datum_yaw);
-    request.geo_pose.orientation = tf2::toMsg(quat);
-    robot_localization::srv::set_datum::response response;
+    request->geo_pose.orientation = tf2::toMsg(quat);
+    auto response = std::make_shared<robot_localization::srv::SetDatum::Response>();
     datumCallback(request, response);
   }
 
-  auto odom_sub = node_->create_subscription<nav_msgs::msg::Odometry>(
-    "odometry/filtered", std::bind(&NavSatTransform::odomCallback, this, _1));
-  auto gps_sub = node_->create_subscription<sensor_msgs::msg::NavSatFix>(
-    "gps/fix", std::bind(&NavSatTransform::gpsFixCallback, this, _1));
+  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+    "odometry/filtered", 1, std::bind(&NavSatTransform::odomCallback, this, _1));
+
+  gps_sub_ = this->create_subscription<sensor_msgs::msg::NavSatFix>(
+    "gps/fix", 1, std::bind(&NavSatTransform::gpsFixCallback, this, _1));
 
   if (!use_odometry_yaw_ && !use_manual_datum_) {
-    auto imu_sub = node_->create_subscription<sensor_msgs::msg::Imu>(
-      "imu", std::bind(&NavSatTransform::imuCallback, this, _1));
+    imu_sub_ = this->create_subscription<sensor_msgs::msg::Imu>(
+      "imu", 1, std::bind(&NavSatTransform::imuCallback, this, _1));
   }
 
-  auto gps_odom_pub =
-    node_->create_publisher<nav_msgs::msg::Odometry>("odometry/gps");
-  rclcpp::Subscription::SharedPtr filtered_gps_pub;
+  gps_odom_pub_ =
+    this->create_publisher<nav_msgs::msg::Odometry>("odometry/gps", rclcpp::QoS(10));
 
   if (publish_gps_) {
-    filtered_gps_pub =
-      node_->create_publisher<sensor_msgs::msg::NavSatFix>("gps/filtered");
+    filtered_gps_pub_ =
+      this->create_publisher<sensor_msgs::msg::NavSatFix>("gps/filtered", rclcpp::QoS(10));
   }
 
   // Sleep for the parameterized amount of time, to give
   // other nodes time to start up (not always necessary)
-  rclcpp::utilities::sleep_for(std::chrono::seconds(delay));
+  rclcpp::sleep_for(std::chrono::duration_cast<std::chrono::seconds>(std::chrono::duration<double>(delay)));
 
-  rclcpp::rate::Rate rate(frequency);
-  while (rclcpp::ok()) {
-    rclcpp::spin_some(node_);
+  int interval = static_cast<int>(1.0/frequency);  
```

Comment:
setting interval as an int here will round it to zero for all frequencies > 1

I think a better type is   
`auto interval = std::chrono::duration<double>(1.0/frequency)
`

and then the wall timer declaration becomes

  `timer_ = this->create_wall_timer(interval, std::bind(&NavSatTransform::transformCallback, this));
`

Reply:
oh nice catch, thank you! will change,